django/middleware/csrf.py

```diff
@@ -160,21 +161,51 @@ def _reject(self, request, reason):
         )
         return _get_failure_view()(request, reason=reason)
 
+    def _get_current_token(self, request):
+        try:
+            if settings.CSRF_USE_SESSIONS:
+                try:
+                    csrf_token = _sanitize_token(request.session[settings.CSRF_SESSION_KEY])
+                except AttributeError:
+                    raise ImproperlyConfigured(
+                        'CSRF_USE_SESSIONS is enabled, but request.session is not set. Please make '
+                        'sure you have sessions set up correctly.'
+                    )
+            else:
+                cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
+                csrf_token = _sanitize_token(cookie_token)
+
+                if csrf_token != cookie_token:
+                    # Cookie token needed to be replaced;
+                    # the cookie needs to be reset.
+                    request.csrf_cookie_needs_reset = True
+            # Use same token next time.
+            request.META['CSRF_COOKIE'] = csrf_token
+            return csrf_token
+        except KeyError:
+            return None
+
+    def _set_token(self, request, response):
+        if settings.CSRF_USE_SESSIONS:
+            request.session[settings.CSRF_SESSION_KEY] = request.META['CSRF_COOKIE']
+        else:
+            response.set_cookie(settings.CSRF_COOKIE_NAME,
+                                request.META['CSRF_COOKIE'],
+                                max_age=settings.CSRF_COOKIE_AGE,
+                                domain=settings.CSRF_COOKIE_DOMAIN,
+                                path=settings.CSRF_COOKIE_PATH,
+                                secure=settings.CSRF_COOKIE_SECURE,
+                                httponly=settings.CSRF_COOKIE_HTTPONLY
+                                )
+            # Content varies with the CSRF cookie, so set the Vary header.
+            patch_vary_headers(response, ('Cookie',))
+
     def process_view(self, request, callback, callback_args, callback_kwargs):
         if getattr(request, 'csrf_processing_done', False):
             return None
 
-        try:
-            cookie_token = request.COOKIES[settings.CSRF_COOKIE_NAME]
-        except KeyError:
-            csrf_token = None
-        else:
-            csrf_token = _sanitize_token(cookie_token)
-            if csrf_token != cookie_token:
-                # Cookie token needed to be replaced;
-                # the cookie needs to be reset.
-                request.csrf_cookie_needs_reset = True
-            # Use same token next time.
+        csrf_token = self._get_current_token(request)
+        if csrf_token is not None:
```

Comment:
Oh, no, this should be finde, I was just wrong in my reasons regarding the sanitization. ``_get_current_token`` actually returns ``None`` if the code does not exist. Taking this into account, I removed the sanitization (because an empty string value shouldn't ever happen).